### PR TITLE
add build task

### PR DESCRIPTION
**Pull request by montague**
Add new `build` task to address #120.

Refactor `build` and `publish` tasks to use new `Mix.Hex.Build` module.

### CODE REVIEW

**Comment by montague** [comment]
would love any feedback on this @ericmj @josevalim @lexmag

File: lib/mix/hex/utils.ex [hunk 1]
```diff
@@ -1,4 +1,4 @@
-defmodule Mix.Tasks.Hex.Utils do
+defmodule Mix.Hex.Utils do
```
Comment (lexmag) [inline lib/mix/hex/utils.ex]: :+1: could it be in separate commit though?

**Comment by lexmag** [comment]
It seems possible to avoid code duplication in `hex.publish` and `hex.build` tasks by running `build` task inside `publish` one.

**Comment by montague** [comment]
good point. i will look into that tomorrow.

File: lib/mix/hex/utils.ex (hunk 1 above)
Comment (montague) [inline lib/mix/hex/utils.ex]: to confirm--you're saying the addition of that file should be in its own commit? sure.

Comment (lexmag) [inline lib/mix/hex/utils.ex]: Yes, this improvement is clearly not related to the main work of PR. It will make easier to reason about in git history. I'd say it can even go into another PR, to reduce changes for review in the current.
As many git-style guides says:
> Don't make several logical changes in one commit.

Comment (montague) [inline lib/mix/hex/utils.ex]: makes sense, thanks for clarifying.

**Comment by ericmj** [comment]
Please rebase and force push instead of closing the PR.

**Comment by montague** [comment]
will do--finishing up the changes soon.

File: lib/mix/hex/build.ex [hunk 2]
```diff
@@ -0,0 +1,210 @@
+defmodule Mix.Hex.Build do
+  @default_files ~w(lib priv mix.exs README* readme* LICENSE*
+                    license* CHANGELOG* changelog* src)
+
+  @error_fields ~w(files app name version build_tools)a
+  @warn_fields ~w(description licenses contributors links)a
+  @meta_fields @error_fields ++ @warn_fields ++ ~w(elixir)a
+
+  def prepare_package! do
```
Comment (montague) [inline lib/mix/hex/build.ex]: I added this to further reduce duplication between `publish` and `build`. I think being able to simply call the `build` task inside the `publish` task (and then push to hex web) would require significantly more refactoring. For example, publishing and reverting both require access to the meta information. If we turn `--revert` into its own mix task, I think it'd be more straightforward to call the `build` task inside `publish`.

File: lib/mix/tasks/hex/publish.ex [hunk 3]
```diff
@@ -76,41 +77,22 @@ defmodule Mix.Tasks.Hex.Publish do
...
-           |> Map.merge(package)
-           |> package(config)
-           |> Map.put(:requirements, deps)
-
-    if Mix.Project.umbrella?(config) do
-      Mix.raise "Hex does not support umbrella projects"
-    end
-
     if version = opts[:revert] do
       revert(meta, version, auth)
```
Comment (montague) [inline lib/mix/tasks/hex/publish.ex]: I think pulling this into its own mix task would make it easier to refactor the `publish` task to simply call `build` and then push to hex web.

**Comment by montague** [comment]
I reverted the namespace change of `Mix.Tasks.Hex.Utils` to avoid cluttering up this PR. I'll reimplement that in a follow-up PR when this one gets merged.

**Comment by montague** [comment]
As always, thanks everyone for taking a look!

File: lib/mix/tasks/hex/build.ex [hunk 4]
```diff
@@ -0,0 +1,77 @@
+defmodule Mix.Tasks.Hex.Build do
+  use Mix.Task
+  alias Mix.Tasks.Hex.Utils
+  alias Mix.Hex.Build
+
+  @shortdoc "Build a new package version"
+
+  @moduledoc """
+  Build a new version of your package.
```
Comment (ericmj) [inline lib/mix/tasks/hex/build.ex]: Can we clarify what this does in another paragraph? `mix compile` also builds the package so it's not clear what this actually does.

Mention that it works like `hex.publish` but it only builds the package tarball into the local directory, what name it will have, and that it doesn't actually publish to a repository.

**Comment by ericmj** [comment]
Looks good. I dropped a small comment where I think documentation can be improved.

**Comment by montague** [comment]
Docs updated

**Comment by ericmj** [comment]
Thank you!
